### PR TITLE
Support PBKDF2 using SHA512

### PR DESCRIPTION
**Pull request by jefftrudeau**

### CODE REVIEW

**Comment by dblessing** [comment]
@carlosantoniodasilva Is this something you're open to accepting? Happy to help move it forward if it's agreeable.

**Comment by dblessing** [comment]
@carlosantoniodasilva Is this something you're open to accepting? Happy to help move it forward if it's agreeable.

File: lib/devise/encryptable/encryptors/pbkdf2.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    module Encryptable
+      module Encryptors
+        class Pbkdf2 < Base
+          def self.compare(encrypted_password, password, stretches, salt, pepper)
+            value_to_test = self.digest(password, stretches, salt, pepper)
+            ActiveSupport::SecurityUtils.fixed_length_secure_compare(encrypted_password, value_to_test)
+          end
+
+          def self.digest(password, stretches, salt, pepper)
+            hash = OpenSSL::Digest::SHA512.new
```
Comment (dblessing) [inline lib/devise/encryptable/encryptors/pbkdf2.rb:12]:
```suggestion
            hash = OpenSSL::Digest.new('SHA512').new
```

File: lib/devise/encryptable/encryptors/pbkdf2.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+begin
+  module Devise
+    module Encryptable
+      module Encryptors
+        class Pbkdf2 < Base
+          def self.compare(encrypted_password, password, stretches, salt, pepper)
+            value_to_test = self.digest(password, stretches, salt, pepper)
+            ActiveSupport::SecurityUtils.fixed_length_secure_compare(encrypted_password, value_to_test)
```
Comment (dblessing) [inline lib/devise/encryptable/encryptors/pbkdf2.rb:8]: If comparing unmatching values of differing length, this fails with mismatch length error. This should use `Devise.secure_compare` instead, which should also do a fixed length comparison, but also does an initial byte size/length check. 


```suggestion
            Devise.secure_compare(encrypted_password, value_to_test)
```

File: lib/devise/encryptable/encryptors/pbkdf2.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+          def self.digest(password, stretches, salt, pepper)
+            hash = OpenSSL::Digest::SHA512.new
+            OpenSSL::KDF.pbkdf2_hmac(
+              password,
+              salt: "#{[salt].pack('H*')}#{pepper}",
+              iterations: stretches,
+              hash: hash,
+              length: hash.digest_length,
+            ).unpack('H*')[0]
```
Comment (dblessing) [inline lib/devise/encryptable/encryptors/pbkdf2.rb:19]:
```suggestion
            ).unpack1('H*')
```

File: lib/devise/encryptable/encryptors/pbkdf2.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        class Pbkdf2 < Base
+          def self.compare(encrypted_password, password, stretches, salt, pepper)
+            value_to_test = self.digest(password, stretches, salt, pepper)
+            ActiveSupport::SecurityUtils.fixed_length_secure_compare(encrypted_password, value_to_test)
+          end
+
+          def self.digest(password, stretches, salt, pepper)
+            hash = OpenSSL::Digest::SHA512.new
+            OpenSSL::KDF.pbkdf2_hmac(
+              password,
```
Comment (dblessing) [inline lib/devise/encryptable/encryptors/pbkdf2.rb:14]: If a password is entirely numeric this function will fail trying due to implicit conversion of integer to string. 

```suggestion
              password.to_s,
```

**Comment by jefftrudeau** [comment]
@dblessing @carlosantoniodasilva I rebased our branch against main and created a new PR: https://github.com/heartcombo/devise-encryptable/pull/25

**Pull request closed without merging**